sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -395,42 +404,50 @@ object FileStreamSource {
     }
 
     override def clean(entry: FileEntry): Unit = {
-      val curPath = new Path(new URI(entry.path))
-      val newPath = new Path(baseArchivePath.toString.stripSuffix("/") + curPath.toUri.getPath)
-
-      try {
-        logDebug(s"Creating directory if it doesn't exist ${newPath.getParent}")
-        if (!fileSystem.exists(newPath.getParent)) {
-          fileSystem.mkdirs(newPath.getParent)
-        }
+      cleanThreadPool.submit(new Runnable {
```

Comment:
You can avoid this by having `clean` be implemented in the base class and have concrete classes implement a `cleanTask` method, for example.

Then it even makes it easy to disable the thread pool altogether (numThreads = 0).